tester/src/test/scala/spinal/tester/code/FormalApi.scala

```diff
@@ -0,0 +1,418 @@
+      if (!doFail) {
+        assume(clockDomain.isResetActive)
+      } else {
+        assumeInitial(clockDomain.isResetActive)
+      }
+      assert(False)
+    }.setDefinitionName("AssertUnderResetTest"))
+  }
+
+  test("FormalApiTest.assertUnderReset.symbiyosys") {
+    runAssertUnderResetTest(symbiYosys)
+  }
+  test("FormalApiTest.assertUnderReset.symbiyosys.fail") {
+    shouldFailWithOutput("Checking assertions in step 1..")( // must not fail in first clock
+      runAssertUnderResetTest(symbiYosys, doFail = true)
+    )
+  }
+  test("FormalApiTest.assertUnderReset.ghdl") { // fails
+    runAssertUnderResetTest(ghdl)
+  }
+  test("FormalApiTest.assertUnderReset.ghdl.fail") {
+    shouldFailWithOutput("Checking assertions in step 1..")( // must not fail in first clock
+      runAssertUnderResetTest(ghdl, doFail = true)
+    )
+  }
+
+  /** Test the generation and verification of assert statement outside the
+    * usual (a)synchronous reset condition.
+    * 
+    * The test works by always (even during reset) asserting on false, which
+    * should obviously immediately trigger. It then assumes the reset to be
+    * always active. If implementation disables the assert during resets it will
+    * not raise exception. Hence an implementation that raises an exception is
+    * passing.
+    * 
+    * @note Test has no failing case, raising an exception (assert triggered) is
+    *       passing the test.
+    *
+    * @param backend formal backend to use, i.e. use SymbiYosys or GHDL
+    */
+  def runAssertWithoutResetTest(backend: SpinalFormalBackendSel): Unit = {
+    val config = getConfig(backend).withAsync.withBMC(3)
+
+    config.doVerify(new Component {
+      assume(clockDomain.isResetActive)
+      clockDomain.withoutReset() {
+        assert(False)
+      }
+    }.setDefinitionName("AssertWithoutResetTest"))
+  }
+
+  test("FormalApiTest.assertWithoutReset.symbiyosys.fail") {
+    shouldFailWithOutput("Assert failed in AssertWithoutResetTest") {
+      runAssertWithoutResetTest(symbiYosys)
+    }
+  }
+  test("FormalApiTest.assertWithoutReset.ghdl.fail") {
+    shouldFailWithOutput("Assert failed in AssertWithoutResetTest") {
+      runAssertWithoutResetTest(ghdl)
+    }
+  }
+
+  /** Test the generation and verification of assert statement inside
+    * conditional scope.
+    * 
+    * This test works by wrapping an always failing assert into a when
+    * conditional scope. In passing test, the when condition is always false,
+    * hence the failing assert should not trigger. The condition of the failing
+    * case is true and the assert should trigger.
+    *
+    * @param backend formal backend to use, i.e. use SymbiYosys or GHDL
+    * @param doFail  false = condition is always false; true = always true
+    */
+  def runAssertWithinCondTest(backend: SpinalFormalBackendSel, doFail: Boolean = false): Unit = {
+    val config = getConfig(backend).withAsync.withBMC(3)
+
+    config.doVerify(new Component {
```

Comment:
In verilog verification mode, the FormalDut is preferred to wrap around the dut.
Where the sync reset maner is guaranteed for formal verification code, but remains the design unchanged.

Reply:
Ah, I see now how your changes in #1435 come together. But still, wrapping it in a `FormalDut` is a bit hideous in this instance. As `FormalDut` and the new `FormalPhase` only work with a hierarchy of components I'd have to write something hideous like: 
```scala
config.doVerify(new Component { FormalDut( new Component {
...
```
This is of course only because the components here are really small and simple. I think I'll just call `setFormalTester()` inside the component and then it's fine.

Reply:
Ok, one thing need to notify is that the way to design the FormalDut is to split the formal verification code a part from the DUT explicitly.

Reply:
Agreed. If the test (i.e. component) would be more complicated I'd also use the FormalDut to split the parts explicitly. For these very simple cases I don't.